### ISSUE TITLE
Mouse & Joystick input not clamping correctly

### ISSUE THREAD

**Issue by jeremyd4500** [godotengine/godot#80534]
### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

M1 Pro MacBook Pro - MacOS Ventura 13.5 & Windows 11

### Issue description

I am experiencing the mouse and joystick input falling out of clamped range when looking/aiming all the way up or down (trying to break out of the vertical clamp range) AND aiming left or right at the same time. For example, if the clamped range for `camera.rotation.x` is (-80, 80) degrees, maxing out the camera's `x` rotation in either of those directions AND moving the camera left or right (essentially moving the camera diagonally) will cause the camera to break out of its clamped `x` rotation and start circling all the way around vertically, at which point the camera's relative transform becomes messed up and starts rotating unexpectedly. Sometimes a strong vertical input from my mouse is enough to break out of the clamped range. 

I have experienced this issue on my MacBook Pro as well as a Windows 11 PC, both with the exact same Godot version.

### Steps to reproduce

The included GitHub tutorial project is a good example. But the important code is in `Scripts/Player.gd` at line 42. 

```
func _unhandled_input(event):
	if event is InputEventMouseMotion:
		head.rotate_y(-event.relative.x * SENSITIVITY)
		camera.rotate_x(-event.relative.y * SENSITIVITY)
		camera.rotation.x = clamp(camera.rotation.x, deg_to_rad(-40), deg_to_rad(60))
```

### Minimal reproduction project

This isn't exactly a "minimal" project, but it is a light-weight tutorial project I found on YouTube. My personal code for handling mouse and joystick aiming is very similar to this project's. And I experienced the exact same issue with this tutorial project:

https://github.com/LegionGames/ShootingMechanics/tree/main

The YouTube video for reference: https://youtu.be/6bbPHsB9TtI

**Comment by Calinou** [comment]
I'd recommend only using the `rotation` property with `+=` instead of using the `rotate_x()` and `rotate_y()` methods.

**Comment by jeremyd4500** [comment]
Thank you for your reply! I will give your suggestion a try. And for my knowledge, is there documentation that I can reference that reflects what you are saying to do?

**Comment by Calinou** [comment]
> And for my knowledge, is there documentation that I can reference that reflects what you are saying to do?

Not that I know of, but the documentation should be updated to mention you shouldn't use the `rotate_*` methods and the `rotation` property in the same frame.

**Comment by jeremyd4500** [comment]
> you shouldn't use the `rotate_*` methods and the `rotation` property in the same frame.

Does this suggest that using the `rotate_*` functions and `camera.rotation.x = ...` together is likely the cause of the issue I'm experiencing?

**Comment by Calinou** [comment]
> Does this suggest that using the rotate_* functions and camera.rotation.x = ... together is likely the cause of the issue I'm experiencing?

Yes, this seems like the most likely cause. The `rotate_*` methods are pretty much legacy at this point – they date back to Godot 1.x back when there was no property syntax for setting member variables.

**Comment by TAGames** [comment]
So far I always used `rotate_*`, but now clamping doesn't work anymore.
Adding to `rotation += ` like Calinou recommends works like a charm.

**Comment by babyl05** [comment]
oh my god i spent so much time on this and this is what worked. thank u thank u thank u

**Comment by Calinou** [comment]
> @Calinou Why? What's the reasoning behind this statement? I don't see anything wrong with doing so.

I agree it should be fixed, but I wouldn't be surprised the current behavior is (accidentally) relied upon in some projects. Fixing the bug would effectively break behavior in projects that rely on the current bug. Given the nature of this change, it might be quite difficult for users to track down.
